build/build/src/ide/web/download.rs

```diff
@@ -0,0 +1,103 @@
+//! Support for downloading files.
+
+use crate::prelude::*;
+
+use ide_ci::cache::Cache;
+use ide_ci::cache::Storable;
+use reqwest::IntoUrl;
+use std::fs::File;
+
+
+
+// ===========================
+// === Download with cache ===
+// ===========================
+
+/// Check whether the specified URL has been downloaded and cached. If not, download it using the
+/// specified downloader, and store it in the cache. Opens the file in the cache and returns it.
+/// The `filename` parameter may be any path-safe string; it will only be seen when inspecting the
+/// cache.
+pub async fn get_file_from_cache_or_download(
+    filename: impl AsRef<Path>,
+    cache: &Cache,
+    downloader: impl Downloader,
+    url: impl IntoUrl,
+) -> Result<File> {
+    let url = url.into_url()?;
+    let filename = filename.as_ref().into();
+    let downloader = Arc::new(downloader);
+    let cache_path = cache.get(FixedContentUrl { url, filename, downloader }).await?;
+    Ok(File::open(&cache_path)?)
+}
+
+/// A URL that can provide a file which can be assumed to be constant.
+#[derive(Debug)]
+struct FixedContentUrl<D> {
+    url:        Url,
+    filename:   Box<Path>,
```

Comment:
Out of curiosity, why `Box<Path>` is used everywhere instead of `PathBuf`?

Reply:
It's a more specific type. `PathBuf` supports all the operations of `Box<Path>`, and additional operations. The additional operations offered by `PathBuf` are not appropriate for this value--it's the file part of a path, so e.g. it makes no sense to append anything to it. This is not a very important distinction in this case, but it also doesn't cost us anything here to be precise.